### PR TITLE
Submesoscale optimisation

### PR DESCRIPTION
**Pull request by russfiedler**
Optimisations for ocean_submesoscale.F90

Contributions are zero below the surface layer  so rather than going to the bottom only perform work where it is needed.
Loops are reorganised to usual order improving performance.
Eliminated unnecessary domain updates.

Reproduces om3_core3 with fp-precise
Approx 30% performance improvement on test case.

### CODE REVIEW

**Comment by aidanheerdegen** [comment]
Nice work @russfiedler. Is this the one that wasn't reproducing after coupling? If so, what did  you have to do to fix that?

**Comment by russfiedler** [comment]
I was just a bit too aggressive with cutting out work to be done by the looks. I'm not totally sure why I couldn't do what I wanted to do. Full 3d dumps at double precision didn't give a hint. It hardly made a difference in timing so I just let it go.

File: src/mom5/ocean_param/lateral/ocean_submesoscale.F90 [hunk 1]
```diff
@@ -1397,23 +1397,27 @@ subroutine compute_bldepth(Time, Thickness, Dens, T_prog, surf_blthick)
...
-                   buoy_freq_ave(i,j) = buoy_freq_ave(i,j) - grav_rho0r*Thickness%dzt(i,j,k)*Dens%drhodz_zt(i,j,k)  
-                enddo
-                buoy_freq_ave(i,j)    = sqrt(abs(buoy_freq_ave(i,j))/mld_thickness) 
-                front_length_inv(i,j) = min(front_length_const_inv, coriolis_param(i,j)/(epsln+mld_thickness*buoy_freq_ave(i,j))) 
-            endif 
-         enddo
-      enddo
+     front_length_inv = front_length_const_inv 
+     wrk1_2d(:,:)         = epsln
+     where(kblt>=min_kblt)
```
Comment (aidanheerdegen) [inline src/mom5/ocean_param/lateral/ocean_submesoscale.F90]: I love a good where more than most, but isn't it mostly redundant with the test at 1410, and the where at 1417?

File: src/mom5/ocean_param/lateral/ocean_submesoscale.F90 [hunk 2]
```diff
@@ -2470,10 +2474,11 @@ subroutine compute_flux_x(Time,n,Tracer)
   integer :: ip, kr
   real    :: tensor_13(isd:ied,jsd:jed,0:1)
   real    :: sumz(isd:ied,jsd:jed,0:1)
-
   flux_x = 0.0
 
-  do k=1,nk
+
+  do k=1,min(maxval(kblt(isc-1:iec,jsc:jec)),nk) ! Only need to compute in surface layer but it will not reproduce for some reason
```
Comment (aidanheerdegen) [inline src/mom5/ocean_param/lateral/ocean_submesoscale.F90]: Is this comment redundant now? It is reproducing right?

File: src/mom5/ocean_param/lateral/ocean_submesoscale.F90 [hunk 3]
```diff
@@ -2587,7 +2592,7 @@ subroutine compute_flux_y(Time,n,Tracer)
 
   flux_y = 0.0
 
-  do k=1,nk
+  do k=1,min(maxval(kblt(isc:iec,jsc-1:jec))+1,nk) ! Only need to comput in surface layer
```
Comment (aidanheerdegen) [inline src/mom5/ocean_param/lateral/ocean_submesoscale.F90]: Do you care about typos in comments? If so, "comput". Also see below

File: src/mom5/ocean_param/lateral/ocean_submesoscale.F90 [hunk 4]
```diff
@@ -1397,23 +1397,27 @@ subroutine compute_bldepth(Time, Thickness, Dens, T_prog, surf_blthick)
 
   ! inverse front length = f/(<N> H), with H=hblt and <N> ave buoyancy freq over hblt
   if(front_length_deform_radius) then 
-      do j=jsd,jed
-         do i=isd,ied
-            buoy_freq_ave(i,j)    = 0.0
-            mld_thickness         = epsln
```
[the remark this answers is not in the document]
Comment (russfiedler) [reply src/mom5/ocean_param/lateral/ocean_submesoscale.F90]: Yes

File: src/mom5/ocean_param/lateral/ocean_submesoscale.F90 (hunk 1 above)
Comment (russfiedler) [reply src/mom5/ocean_param/lateral/ocean_submesoscale.F90]: It's needed for reproducibility with older results and the diagnostic. You need to initialise `buoy_freq_ave` to `epsln` where you are going to do the averaging and zero elsewhere. As you point out the division by zero issue is already avoided.

File: src/mom5/ocean_param/lateral/ocean_submesoscale.F90 (hunk 2 above)
Comment (russfiedler) [reply src/mom5/ocean_param/lateral/ocean_submesoscale.F90]: Correct

**Comment by aidanheerdegen** [comment]
One last question @russfiedler and then I'm happy to merge this ... would you achieve the same speed up by keeping the looping as it was and `break`ing out of the `do k=1,kblt(i,j)` loop when you hit `maxval(kblt)`?

**Comment by russfiedler** [comment]
I think it's a bit difficult to break out of loop which can't reach a value. By definition `kbl(i,j) <= maxval(kbl)`!
The main advantage of the loop reordering is memory access albeit at the cost of an `if` test. A 3D mask for `kblt(i,j)>=max(min_kbl,k)` could be constructed in order to avoid that but I'm not sure there would be much benefit.

**Comment by aidanheerdegen** [comment]
Heh. Well now I'm questioning the value of me reviewing this when I couldn't even figure that out. But thanks for enlightening me. I'll merge now.

**Pull request merged**